HashAlgorithmName one-shots

vcsjones opened the pull request:
This implements CryptographicOperations.{HashData,HmacData} and removes our internal helper in favor of using the public one.

Closes #91407.

vcsjones committed:
Implement HMAC and hash one-shot for HashAlgorithmName.

vcsjones commented on src/libraries/System.Security.Cryptography/tests/HmacMD5Tests.cs:
```diff
@@ -15,9 +15,11 @@ public class HmacMD5Tests : Rfc2202HmacTests<HmacMD5Tests.Traits>
         {
             public static bool IsSupported => true;
-            public static int HashSizeInBytes => HMACSHA1.HashSizeInBytes;
+            public static int HashSizeInBytes => HMACMD5.HashSizeInBytes;
```
This turned out to be a mistake in our HMACMD5 tests but it went unnoticed because the value is only used for "create a buffer that is at _least_ large enough". LEN(SHA1) > LEN(MD5) so it happened to work.  But let's fix it.

vcsjones commented on src/libraries/System.Security.Cryptography/tests/HmacTests.cs:
The `HmacTests` are in desperate need of some refactoring and love in my opinion. I will look at that hopefully sometime for .NET 9. I would like to re-model the tests similarly to how we did SHAKE.

For now though, I decided to keep the changes small so as to not make too many test changes at the same time as product changes.

vcsjones commented on src/libraries/System.Security.Cryptography/tests/HashAlgorithmTestDriver.cs:
Though the implementation is on `CryptographicOperations`, the tests are all part of the hashing / HMAC testing infrastructure. Otherwise we would be duplicating a lot of additional test logic, KATs, etc.

vcsjones commented:
These native AOT outerloop checks looks like they ran errantly according to https://github.com/dotnet/runtime/issues/92432, so though they are failing, I marked this ready for review.

stephentoub commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/CryptographicOperations.cs:
```diff
@@ -58,5 +64,766 @@ public static void ZeroMemory(Span<byte> buffer)
+        {
+            if (!TryHashData(hashAlgorithm, source, destination, out int bytesWritten))
+            {
+                throw new ArgumentException(SR.Argument_DestinationTooShort, nameof(destination));
```
This shows up at least 6 times. Might be worth creating a private static throw helper method for it, both to reduce the duplication and increase the liklihood this will be inlined.

vcsjones replied:
There are some subtle differences between these. Some are on ROS, some are on ROM. That would mean using `.Span` on the memory, which can be a little expensive. Or we have two helpers, which I am not sure helps at that point.

Unless you meant for the helper to be

```C#
private static void CheckDestinationSize(int requiredSize, int destinationSize)
{
    if (destinationSize < requiredSize)
    {
        throw new ArgumentException(SR.Argument_DestinationTooShort, "destination");
    }
}

// Used like

CheckDestinationSize(hashLength, destination.Length);
```

stephentoub replied:
I actually just meant:
```C#
private static void ThrowDestinationTooShort() => throw new ArgumentException(SR.Argument_DestinationTooShort, "destination");
```
but we could do what you suggest, too.

vcsjones committed:
Use throw helper for destination length.

Also change the variable name for the hash length to be consistent accross all methods.

stephentoub commented:
Thanks.
